src/caches/abstractbeliefpropagationcache.jl

```diff
@@ -24,26 +25,48 @@ function data_graph_type(bpc::AbstractBeliefPropagationCache)
 end
 data_graph(bpc::AbstractBeliefPropagationCache) = data_graph(tensornetwork(bpc))
 
-function default_message_update(contract_list::Vector{ITensor}; normalize=true, kwargs...)
-  sequence = contraction_sequence(contract_list; alg="optimal")
-  updated_messages = contract(contract_list; sequence, kwargs...)
+function message_update_function(
+  alg::Algorithm"contract",
+  contract_list::Vector{ITensor};
+  normalize=alg.kwargs.normalize,
+  sequence_alg=alg.kwargs.sequence_alg,
+)
+  sequence = contraction_sequence(contract_list; alg=sequence_alg)
+  updated_messages = contract(contract_list; sequence)
   message_norm = norm(updated_messages)
   if normalize && !iszero(message_norm)
     updated_messages /= message_norm
   end
   return ITensor[updated_messages]
 end
 
+function message_update_function(
+  alg::Algorithm"contract_custom_device",
```

Comment:
I think we should make this more general and not specific to using "contract" as an update. For example, we could call it `Algorithm"adapt_update"` and expect it to have keyword arguments `alg.kwargs.adapt` and `alg.kwargs.alg`, where `alg.kwargs.alg` gets passed as the algorithm to `message_update_function` once the messages are adapted.

If users are using "contract" a lot I think they could make a more convenient wrapper if they want that for convenience, but I think writing:
```julia
update(bpc; message_update_alg=Algorithm("adapt_update"; adapt=CUDA.cu, alg=Algorithm("contract"; normalize=true, sequence_alg = "optimal")))
```
isn't so bad and makes it clearer to users what is going on and how they could customize the behavior.